src/lib/reasoners/satml_frontend.ml

```diff
@@ -1035,6 +1038,117 @@ module Make (Th : Theory.S) : Sat_solver_sig.S = struct
+     other branch of the SAT solver in which all the formulas are satisfiable
+     and the objective functions have bigger values in some models of this
+     branch.
+
+     For instance, consider the SMT-LIB problem:
+       (declare-const x Int)
+       (assert (or (<= x 2) (<= x 4)))
+       (assert (or (<= y 3))
+       (maximize x)
+       (maximize y)
+       (check-sat)
+       (get-model)
+
+     Assume that the solver explores the first branch (<= x 2) of the or gate:
+       (or (<= x 2) (<= x 4)).
+
+     Let's imagine it discovers that [0] is a model of the first formula. After
+     optimization, it find that [2] is the best model for [x] and [3] is the
+     best model for [y] it can got in this branch. Still we have to explore
+     the second branch [(<= x 4)] to realize that [4] is actually the best
+     model for [x].
+
+     The following function ensures to explore adequate branches of the SAT
+     solver in order to get the best model, if the problem is SAT.
+
+     In our running example, after getting the model [2], the below function
+     run again the SAT solver with the extra assert:
+       (assert (or (> x 2) (and (= x 2) (> y 3)))
+
+     If this run produces the answer [unsat], we know that [2] was the best
+     model value for [x]. Otherwise, we got a better value for [x]. *)
+
+  let op is_max is_le =
+    if is_max then begin
+      if is_le then Expr.Reals.(<=)
+      else Expr.Reals.(<)
+    end
+    else begin
+      if is_le then Expr.Reals.(>=)
+      else Expr.Reals.(>)
+    end
+
+  let analyze_unknown_for_objectives env unknown_exn unsat_rec_prem : unit =
+    let obj = Th.get_objectives (SAT.current_tbox env.satml) in
+    if Util.MI.is_empty obj then raise unknown_exn;
+    env.objectives := Some obj;
+    let acc =
+      try
+        Util.MI.fold
+          (fun _ {Th_util.e; value; r=_; is_max; _} acc ->
+             match value with
+             | Pinfinity | Minfinity ->
+               raise (Give_up acc)
+             | Value v ->
+               (e, v, is_max, true) :: acc
+             | Limit (_, v) ->
+               raise (Give_up ((e, v, is_max, false) :: acc))
+             | Unknown ->
+               assert false
+          ) obj []
+      with Give_up acc -> acc
+    in
+    begin match acc with
+      | [] ->
+        (* The answer for the first objective is infinity. We stop here as
+           we cannot go beyond infinity and the next objectives with lower
+           priority cannot be optimized in presence of infinity values. *)
+        raise unknown_exn;
+
+      | (e, tv, is_max, is_le) :: l ->
+        let neg =
+          List.fold_left
+            (fun acc (e, tv, is_max, is_le) ->
+               let eq = E.mk_eq e tv ~iff: false in
+               let acc = E.Core.and_ acc eq in
+               E.Core.(or_ acc (not ((op is_max is_le) e tv)))
+            ) (E.Core.not ((op is_max is_le) e tv)) l
```

Comment:
Assuming `is_max` is always `true`:

Inside the `fold_left`, if we encounter a second expression.
 - With `is_le = true`, we have found `e ≤ tv`. We need to look for *either* `e > tv`, or  (due to the lexicographic ordering) `e = tv` and `acc`, which is what we build.
 - With `is_le = false`, we have found `e < tv`. To contradict that, we just need to look for `e ≥ tv` — the values for the other terms cannot be optimized yet (note that this does mean that once we find a "strict" bound the rest is best-effort, which we know already). Currently we look for `e ≥ tv ∨ (e = tv ∧ acc)`, but the second branch is subsumed by the first (`e = tv` implies `e ≥ tv`).

So technically this is correct but we should just return `E.Core.(not ((op is_max is_le) e tv))` when `is_le` is `true` in the accumulator.

Reply:
As it's an optimization, I will apply it in #921 ;)